symengine/basic_to_upoly.cpp

```diff
@@ -0,0 +1,317 @@
+#include <symengine/visitor.h>
+
+namespace SymEngine
+{
+
+class UIntPolyGenaratorVisitor : public BaseVisitor<UIntPolyGenaratorVisitor>
+{
+private:
+    RCP<const Basic> res;
+    bool inpow;
+
+public:
+    RCP<const Basic> apply(const Basic &b, bool inpow_)
+    {
+        inpow = inpow_;
+        b.accept(*this);
+        return res;
+    }
+
+    void bvisit(const Pow &x)
+    {
+        if (is_a<const Rational>(*x.get_base()) or is_a<const Integer>(*x.get_base())) {
+            res = pow(x.get_base(), _find_gen_uintpoly(x.get_exp(), true));
+        } else if (is_a<const Rational>(*x.get_exp()) or is_a<const Integer>(*x.get_exp())) {
+
+            int tmp = positive_integer(x.get_exp());
+            if (tmp != 0) {
+                if (tmp > 0)
+                    res = _find_gen_uintpoly(x.get_base());
+                else
+                    res = x.rcp_from_this();
+                return;
+            }
+            
+            if (is_a<const Rational>(*x.get_exp())) {
+                RCP<const Rational> rat = rcp_static_cast<const Rational>(x.get_exp());
+                if (rat->is_negative())
+                    res = pow(x.get_base(), div(neg(one), rat->get_den()));
+                else
+                    res = pow(x.get_base(), div(one, rat->get_den()));
+            }
+            else {
+                throw std::runtime_error("Could not extract generator");
+            }
+        } else {
+            // won't handle cases like genarator = x**x
+            throw std::runtime_error("Could not extract generator");
+        }
+    }
+
+    void bvisit(const Add &x)
+    {
+        RCP<const Basic> min_gen = null, curr_gen;
+
+        if (not inpow)
+            if (not is_a<const Integer>(*x.coef_))
+                throw std::runtime_error("Non-integer coeff found");
+
+        for (const auto &it : x.dict_) {
+            if (not inpow)
+                if (not is_a<const Integer>(*it.second))
+                    throw std::runtime_error("Non-integer coeff found");
+
+            curr_gen = _find_gen_uintpoly(it.first);
+
+            if(min_gen != null) {
+                if ((is_a_sub<const Function>(*min_gen) and is_a_sub<const Function>(*curr_gen)) or
```

Comment:
No, because say `min_gen` was `x` and `curr_gen` was `x**(1/2)`, they are unequal but we can indeed use `x**(1/2)` as a generator. 

The strucutre is not very clear right now, I will try to make amends.